service/history/shard/fx.go

```diff
@@ -48,48 +48,35 @@ var Module = fx.Options(
-	throttledLogger resource.ThrottledLogger,
-	persistenceExecutionManager persistence.ExecutionManager,
-	persistenceShardManager persistence.ShardManager,
-	clientBean client.Bean,
-	historyClient historyservice.HistoryServiceClient,
-	historyServiceResolver membership.ServiceResolver,
-	metricsClient metrics.Client,
-	payloadSerializer serialization.Serializer,
-	timeSource clock.TimeSource,
-	namespaceRegistry namespace.Registry,
-	saProvider searchattribute.Provider,
-	saMapper searchattribute.Mapper,
-	clusterMetadata cluster.Metadata,
-	archivalMetadata archiver.ArchivalMetadata,
-	hostInfoProvider resource.HostInfoProvider,
-) *ControllerImpl {
+type ShardControllerDeps struct {
+	fx.In
+
+	Config                      *configs.Config
+	Logger                      log.Logger
+	ThrottledLogger             resource.ThrottledLogger
+	PersistenceExecutionManager persistence.ExecutionManager
+	PersistenceShardManager     persistence.ShardManager
+	ClientBean                  client.Bean
+	HistoryClient               historyservice.HistoryServiceClient
+	HistoryServiceResolver      membership.ServiceResolver
+	MetricsClient               metrics.Client
+	PayloadSerializer           serialization.Serializer
+	TimeSource                  clock.TimeSource
+	NamespaceRegistry           namespace.Registry
+	SaProvider                  searchattribute.Provider
+	SaMapper                    searchattribute.Mapper
+	ClusterMetadata             cluster.Metadata
+	ArchivalMetadata            archiver.ArchivalMetadata
+	HostInfoProvider            resource.HostInfoProvider
+}
+
+func ShardControllerProvider(d ShardControllerDeps) *ControllerImpl {
 	return &ControllerImpl{
-		status:                      common.DaemonStatusInitialized,
-		membershipUpdateCh:          make(chan *membership.ChangedEvent, 10),
-		historyShards:               make(map[int32]*ContextImpl),
-		shutdownCh:                  make(chan struct{}),
-		logger:                      logger,
-		contextTaggedLogger:         logger,          // will add tags in Start
-		throttledLogger:             throttledLogger, // will add tags in Start
-		config:                      config,
-		metricsScope:                metricsClient.Scope(metrics.HistoryShardControllerScope),
-		persistenceExecutionManager: persistenceExecutionManager,
-		persistenceShardManager:     persistenceShardManager,
-		clientBean:                  clientBean,
-		historyClient:               historyClient,
-		historyServiceResolver:      historyServiceResolver,
-		metricsClient:               metricsClient,
-		payloadSerializer:           payloadSerializer,
-		timeSource:                  timeSource,
-		namespaceRegistry:           namespaceRegistry,
-		saProvider:                  saProvider,
-		saMapper:                    saMapper,
-		clusterMetadata:             clusterMetadata,
-		archivalMetadata:            archivalMetadata,
-		hostInfoProvider:            hostInfoProvider,
+		d:                  d,
```

Comment:
That's the main thing I was trying to get rid of, though: having to repeat the type and field names of each dependency multiple times. Why do you want to add some repetition back? To add friction for introducing new dependencies? :)

I did realize I could make the deps struct private, although it looks like that still exposes the deps from ContextImpl since they're public within the deps struct. We could make ContextImpl itself private to avoid that, I think.